core/src/blocks/extrinsics.rs

```diff
@@ -2,39 +2,60 @@
 // This file is dual-licensed as Apache-2.0 or GPL-3.0.
 // see LICENSE for license details.
 
+use super::BlockError;
 use crate::blocks::extrinsic_signed_extensions::ExtrinsicSignedExtensions;
 use crate::{
     config::{Config, Hasher},
-    error::{BlockError, Error, MetadataError},
+    error::{Error, MetadataError},
     Metadata,
 };
 use alloc::sync::Arc;
 use alloc::vec::Vec;
-use codec::{Compact, CompactLen, Decode};
+use core::ops::Deref;
+use frame_decode::extrinsics::Extrinsic;
 use scale_decode::DecodeAsType;
 use subxt_metadata::PalletMetadata;
 
 pub use crate::blocks::StaticExtrinsic;
 
 /// The body of a block.
 pub struct Extrinsics<T: Config> {
-    extrinsics: Vec<Vec<u8>>,
+    extrinsics: Vec<Arc<(Extrinsic<'static, u32>, Vec<u8>)>>,
     metadata: Metadata,
-    ids: ExtrinsicPartTypeIds,
     _marker: core::marker::PhantomData<T>,
 }
 
 impl<T: Config> Extrinsics<T> {
     /// Instantiate a new [`Extrinsics`] object, given a vector containing
     /// each extrinsic hash (in the form of bytes) and some metadata that
     /// we'll use to decode them.
-    pub fn decode_from(extrinsics: Vec<Vec<u8>>, metadata: Metadata) -> Result<Self, BlockError> {
-        let ids = ExtrinsicPartTypeIds::new(&metadata)?;
+    pub fn decode_from(extrinsics: Vec<Vec<u8>>, metadata: Metadata) -> Result<Self, Error> {
```

Comment:
I moved the decoding work to this initial step.

This means that on the downside we:
- try to decode everything straight away, even if the user doesn't care (though I suspect they usually would)
- an error decoding an extrinsic will lead to a failure early on, preventing eg getting the extrinsic bytes

On the plus side, we:
- Can iterate over or `find` or whatever multiple times without re-doing the decoding work each time we do these things.